hexrdgui/masking/mask_manager.py

```diff
@@ -0,0 +1,390 @@
+import math
+import numpy as np
+
+from PySide6.QtCore import Signal, QObject
+from hexrdgui import utils
+
+from hexrdgui.constants import ViewType
+from hexrdgui.masking.constants import MaskType
+from hexrdgui.masking.create_polar_mask import (
+    create_polar_mask_from_raw, rebuild_polar_masks
+)
+from hexrdgui.masking.create_raw_mask import (
+    recompute_raw_threshold_mask, create_raw_mask, rebuild_raw_masks
+)
+from hexrdgui.hexrd_config import HexrdConfig
+from hexrdgui.masking.mask_compatability import load_masks_v1_to_v2
+from hexrdgui.singletons import QSingleton
+from hexrdgui.utils import unique_name
+
+from hexrd.instrument import unwrap_dict_to_h5
+
+from abc import ABC, abstractmethod
+
+
+class Mask(ABC):
+    def __init__(self, name='', mtype='', visible=True):
+        self.type = mtype
+        self.name = name
+        self.visible = visible
+        self.masked_arrays = None
+
+    def update_mask_visibility(self, visibility):
+        self.visible = visibility
+
+    # Abstract methods
+    @abstractmethod
+    def get_data(self):
+        pass
+
+    @abstractmethod
+    def set_data(self, data):
+        pass
+
+    @abstractmethod
+    def update_masked_arrays(self):
+        pass
+
+    @abstractmethod
+    def serialize(self):
+        pass
+
+    @classmethod
+    def deserialize(cls, data):
+        return cls(data['name'], data['mtype'], data['visible'])
+
+
+class RegionMask(Mask):
+    def __init__(self, name='', mtype='', visible=True):
+        super().__init__(name, mtype, visible)
+        self._raw = None
+
+    def get_data(self):
+        return self._raw
+
+    def set_data(self, data):
+        self._raw = data
+        self.update_masked_arrays()
+
+    def update_masked_arrays(self, view=ViewType.raw):
+        if view == ViewType.raw:
+            self.masked_arrays = create_raw_mask(self._raw)
+        else:
+            self.masked_arrays = create_polar_mask_from_raw(self._raw)
+
+    def serialize(self):
+        data = {
+            'name': self.name,
+            'mtype': self.type,
+            'visible': self.visible,
+        }
+        for i, (det, values) in enumerate(self._raw):
+            data.setdefault(det, {})[str(i)] = values
```

Comment:
Can we put this under a `data` key instead, so it looks like this:

```python
data = {
    'name': self.name,
    'mtype': self.type,
    'visible': self.visible,
    'data': {
        'detector_1': ...,
        'detector_2': ...,
    },
}
```

Although it isn't likely to happen, this will help us ensure that the user's detector names will not clash with the keywords `visible`, `mtype`, etc. That would be a difficult-to-track bug if it were to occur.